data_juicer/ops/mapper/generate_caption_mapper.py

```diff
@@ -0,0 +1,238 @@
+import copy
+import random
+
+import numpy as np
+from jsonargparse.typing import PositiveInt
+
+from data_juicer.utils.availability_utils import AvailabilityChecking
+from data_juicer.utils.constant import Fields
+from data_juicer.utils.mm_utils import (SpecialTokens,
+                                        insert_texts_after_placeholders,
+                                        load_image, remove_non_special_tokens,
+                                        remove_special_tokens)
+from data_juicer.utils.model_utils import get_model, prepare_model
+
+from ..base_op import OPERATORS, Mapper
+from ..deduplicator.document_simhash_deduplicator import \
+    DocumentSimhashDeduplicator
+from ..op_fusion import LOADED_IMAGES
+
+OP_NAME = 'generate_caption_mapper'
+
+with AvailabilityChecking(['torch', 'transformers'], OP_NAME):
+    import torch
+    import transformers  # noqa: F401
+
+    # avoid hanging when calling blip2 in multiprocessing
+    torch.set_num_threads(1)
+
+
+def jaccard_similarity_np(hash_val1, hash_val2):
+    equal_hashes = np.sum(hash_val1 == hash_val2)
+    total_hashes = len(hash_val1)  # Assume both vectors are the same length
+    return equal_hashes / total_hashes
+
+
+@OPERATORS.register_module(OP_NAME)
+@LOADED_IMAGES.register_module(OP_NAME)
+class GenerateCaptionMapper(Mapper):
+    """Mapper to generate samples whose captions are generated based on
+    another model and the figure."""
+
+    def __init__(self,
+                 hf_blip2='Salesforce/blip2-opt-2.7b',
+                 caption_num: PositiveInt = 1,
+                 keep_candidate_mode: str = 'random_any',
+                 *args,
+                 **kwargs):
+        """
+        Initialization method.
+
+        :param hf_blip2: blip2 model name on huggingface to generate caption
+        :param caption_num: how many candidate captions to generate
+        for each image
+        :param keep_candidate_mode: retain strategy for the generated
+        $caption_num$ candidates.
+            'random_any': Retain the random one from generated captions
+            'similar_one': Retain the generated one that is most similar to the
+                original caption
+            'all': Retain all generated captions by concatenation
+        Note: This is a batched_OP, whose input and output type are
+            both list. Suppose there are $N$ list of input samples, whose batch
+            size is $b$, and denote caption_num as $M$.
+            The number of total samples after generation is $2Nb$
+            for 'random_any' and 'similar_one' mode,
+             and $(1+M)Nb$ for 'all' mode.
+        :param args: extra args
+        :param kwargs: extra args
+        """
+        super().__init__(*args, **kwargs)
+        self.is_batched_op = True
```

Comment:
May be `self._is_batched_op = True`,  because `is_batched_op` is a function in base class Mapper.

Reply:
self._batched_op = True

Ref: https://github.com/alibaba/data-juicer/blob/a3c8310bf0848e787fba5eca2373e69a25767fd7/data_juicer/ops/base_op.py#L54